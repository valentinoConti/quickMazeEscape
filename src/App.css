body {
  display: flex;
  justify-content: center;
  margin: 0;
  overflow: hidden;
}

#root {
  background-image: linear-gradient(#755748, #ffc197);
  position: absolute;
  width: 750px;
  height: 750px;
  transform: scale(0.9);
  transform-origin: top center;
}