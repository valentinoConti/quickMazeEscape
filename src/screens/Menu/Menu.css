#menuScreen {
  position: absolute;
  width: 100%;
  height: 100%;
  background: url('../../../assets/menu.png');
  outline: black 2px solid;
  font-size: 24px;
  text-align: center;
}

footer {
  position: absolute;
  bottom: 10px;
  left: 15px;
  font-size: 24px;
  text-shadow: 2px 2px 4px black;
  color: black;
}

.menuTitle {
  text-align: right;
  margin-right: 100px;
  color: black;
  text-shadow: 1px 1px 2px black;
}

.menuButton {
  margin-left: 280px;
  width: 220px;
  height: 40px;
  background-color: white;
  border: 2px solid black;
  box-shadow: 2px 2px 2px black;
  cursor: pointer;
  font-size: 18px;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
}