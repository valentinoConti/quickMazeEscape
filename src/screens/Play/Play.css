#game {
  position: absolute;
}
#game:fullscreen {
  width: 1024px;
  height: 1024px;
}

.gameFooter {
  display: flex;
  height: 56px;
  border-bottom: 2px solid black;
}

.info {
  width: 33%;
  font-size: 24px;
  padding-top: 14px;
  margin-top: 0;
  margin-bottom: 0;
}

.alignLeft {
  text-align: left;
  margin-left: 10px;
}

.alignMiddle {
  text-align: center;
}

.alignRight {
  text-align: right;
  margin-right: 8px;
}

.exitDoor {
  background-image: url('../../../assets/puerta.png');
  background-size: 100%;
  position: absolute;
}

.player {
  background-image: url('../../../assets/personaje.png');
  background-size: 100%;
  position: absolute;
  z-index: 2;
  transition: left 0.15s linear, bottom 0.15s linear;
}

.unlock {
  background-image: url('../../../assets/bomba.png');
  background-size: 100%;
  position: absolute;
  margin-bottom: 15px;
  margin-left: 10px;
}

.lock {
  background-image: url('../../../assets/ladrillo.gif');
  background-size: 50%;
  position: absolute;
  margin-left: -10px;
}

.lost {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 3;
}
.lost > b {
  color: white;
  font-size: 72px;
  position: absolute;
  top: 150px;
  left: 260px
}
.lost > .menuButton.menu {
  margin-top: 30px;
}
.lost > .menuButton.restart {
  margin-top: 300px;
}