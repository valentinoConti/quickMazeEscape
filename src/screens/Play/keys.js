/**
 * @name handledKeys
 * @description 
 * @param {*} prevState
 * @param {*} boardSize
 * @param {*} setMovesLeft
 * @param {*} locked
 */
export const handleKeys = (prevState, level, setMovesLeft, locked) => {
  const X = prevState[0];
  const Y = prevState[1];
  const boardSize = level.tiles.length;
  return {
    ArrowRight: () => {
      if (X < boardSize - 1 && level.tiles[Y][X+1] === 0 && (!locked || !(level.lock[0] === Y && level.lock[1] === X+1))) {
        setMovesLeft(prev => prev - 1);
        return [X+1, Y, {right: true, position: 'normal'}];
      }
      return prevState;
    },
    ArrowLeft: () => {
      if (X > 0 && level.tiles[Y][X-1] === 0 && (!locked || !(level.lock[0] === Y && level.lock[1] === X-1))) {
        setMovesLeft(prev => prev - 1);
        return [X-1, Y, {right: false, position: 'normal'}];
      }
      return prevState;
    },
    ArrowDown: () => {
      if (Y < boardSize - 1 && level.tiles[Y+1][X] === 0 && (!locked || !(level.lock[0] === Y+1 && level.lock[1] === X))) {
        setMovesLeft(prev => prev - 1);
        return [X, Y+1, ({right: prevState[2].right, position: 'down'})];
      }
      return prevState;
    },
    ArrowUp: () => {
      if (Y > 0 && level.tiles[Y-1][X] === 0 && (!locked || !(level.lock[0] === Y-1 && level.lock[1] === X))) {
        setMovesLeft(prev => prev - 1);
        return [X, Y-1, ({right: prevState[2].right, position: 'up'})];
      }
      return prevState;
    }
  }
};

export const allowedKeys = ['ArrowRight', 'ArrowLeft', 'ArrowDown', 'ArrowUp', 'r', 'R'];