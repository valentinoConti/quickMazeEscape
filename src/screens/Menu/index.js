import React from 'react';
import PropTypes from 'prop-types';
import './Menu.css';

/**
 * @name Menu
 * @description Menu screen. Here we have our different screens links (Play, Instructions and Fullscreen toggle)
 * @param {Function} setScreen 
 */
const Menu = ({setScreen}) => {
  const playPressed = () => setScreen('play');
  const controlsPressed = () => setScreen('controls');
  const toggleFullscreen = () => {
    const doc = document.documentElement;
    if (doc.requestFullscreen) {
      doc.requestFullscreen();

      let scale = 0.9;
      while (
        document.querySelector('#root').getBoundingClientRect().height+75 < window.innerHeight &&
        document.querySelector('#root').getBoundingClientRect().width < window.innerWidth
      ) {
        scale += 0.025;
        document.querySelector('#root').style.transform = `scale(${scale})`;
      }
    }
  };
  
  return (
    <div id="menuScreen">
      <h1 className="menuTitle">Quick-Maze Escape</h1>

      <button
        id="playButton"
        className="menuButton"
        onClick={playPressed}>
        Play
      </button>

      <button
        id="instructionsButton"
        className="menuButton marginTop"
        onClick={controlsPressed}>
        Instructions
      </button>

      <button
        id="fullscreenButton"
        className="menuButton marginTop"
        onClick={toggleFullscreen}>
        Fullscreen
      </button>

      <footer>
        Valentino Conti
      </footer>
    </div>
  );
};

Menu.propTypes = {
  setScreen: PropTypes.func.isRequired
};

export default Menu;
