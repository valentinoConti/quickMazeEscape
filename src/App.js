import React, { useState } from 'react';
import './App.css';
import Menu from './screens/Menu';
import Play from './screens/Play';
import Controls from './screens/Controls';
import Win from './screens/Win';

/**
 * @name App
 * @description Quick Maze Escape main App screen. Responsible of the screens management and all the app flux.
 * 
 * @author Valentino Conti
 */
const App = () => {
  const [screen, setScreen] = useState('menu');
  const [levelNumber, setLevelNumber] = useState(1);
  const [score, setScore] = useState(0);

  const [randomKey, setRandom] = useState(Math.random());
  const nextLevel = (which) => {
    setLevelNumber(which);
    setRandom(Math.random());
  }

  return (
    <>
      {screen === 'menu' && <Menu setScreen={setScreen} />}
      {screen === 'controls' && <Controls setScreen={setScreen} />}
      {screen === 'play' && (
        <Play
          setScreen={setScreen}
          levelNumber={levelNumber}
          setLevelNumber={nextLevel}
          setScore={setScore}
          score={score}
          key={`level-${levelNumber}-${randomKey}`}
        />
      )}
      {screen === 'win' && <Win setScreen={setScreen} score={score} setScore={setScore} />}
    </>
  );
};

export default App;
