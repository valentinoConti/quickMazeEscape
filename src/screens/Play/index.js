import React, { useState, useEffect, useMemo, useCallback, useRef } from 'react';
import PropTypes from 'prop-types';
import getLevel from '../../levels';
import { handleKeys, allowedKeys } from './keys';
import './Play.css';

/**
 * @name Play
 * @description This is the game screen. Here is where all the game logic happens
 *
 * @param {Integer} levelNumber
 * @param {Integer} score
 * @param {Function} setLevelNumber
 * @param {Function} setScore
 * @param {Function} setScreen
 */
const Play = ({
  levelNumber,
  score,
  setLevelNumber,
  setScore,
  setScreen
}) => {
  const [playing, setPlaying] = useState(true);

  const proccesedLevel = useMemo(() => getLevel(levelNumber), [levelNumber]);
  const [level] = useState(proccesedLevel);

  const [player, setPlayer] = useState(level.playerStart.concat({
    right: true,
    position: 'normal'
  }));

  const [locked, setLocked] = useState(true);
  const [timeLeft, setTimeLeft] = useState(level.time);
  const [movesLeft, setMovesLeft] = useState(level.movements);
  const gameDivRef = useRef();

  const boardSize = level.tiles.length;

  const getMapPosition = useCallback((bottomValue, leftValue, sizeMultiplier) => ({
    bottom: 58 + (750 / boardSize) * Math.abs(bottomValue - (boardSize - 1)),
    left: 10 + (750 / boardSize) * leftValue,
    width: sizeMultiplier * 750 / boardSize,
    height: sizeMultiplier * 750 / boardSize
  }), [level]);

  const getPlayerPosition = useCallback(() => ({
    bottom: 75 + (750 / boardSize) * Math.abs(player[1] - (boardSize - 1)),
    left: 15 + (750 / boardSize) * player[0],
    width: 11 / 15 * 750 / boardSize,
    height: 11 / 15 * 750 / boardSize,
    transform: `
      scaleX(${player[2].right ? '1' : '-1'})
      rotate(${player[2].position === 'normal'
        ? '0deg'
        : player[2].position === 'up'
          ? '-90deg'
          : '90deg'
      })
    `
  }), [player]);

  const handleKeydown = (ev) => {
    ev.persist();
    ev.preventDefault();
    if (playing && allowedKeys.includes(ev.key)) {
      if (ev.key.toLowerCase() !== 'r') {
        setPlayer(prevState => {
          return handleKeys(prevState, level, setMovesLeft, locked)[ev.key]()
        });
      }
      else {
        lost();
      }
    }
  };

  const lost = () => {
    setPlaying(false);
  }

  // Timer counter and lost
  useEffect(() => {
    if (playing && timeLeft) {
      var timing = setTimeout(() => {
        setTimeLeft(state => state - 1);
      }, 1000);
    }
    else {
      lost()
    }
    return () => clearTimeout(timing);
  }, [timeLeft]);

  // Movements left lost
  useEffect(() => {
    if (!movesLeft) {
      lost()
    }
  }, [movesLeft]);

  // Hit-Test
  useEffect(() => {
    // WIN
    if (player[0] === level.escapeDoor[0] && player[1] === level.escapeDoor[1]) {
      setScore(score + (timeLeft * 3 + movesLeft * 4));
      if (levelNumber < document.totalLevels) {
        setLevelNumber(levelNumber + 1);
      } else {
        setLevelNumber(1);
        setScreen('win');
      }
    }
    // bomb unlock
    else if (player[0] === level.unlock[1] && player[1] === level.unlock[0]) {
      setLocked(false);
    }
  }, [player]);

  useEffect(() => { gameDivRef.current?.focus() }, []);
  const goMenu = () => {
    setScreen('menu')
    setScore(0);
    setLevelNumber(1);
  }

  const restart = () => {
    setScreen('play')
    setScore(0);
    setLevelNumber(1);
  }

  return (
    <div id="game" tabIndex="0" onKeyDown={handleKeydown} ref={gameDivRef}>
      {!playing && (
        <div className="lost">
          <b>You lose</b>
          <button className="menuButton restart" onClick={restart}>Restart</button>
          <button className="menuButton menu" onClick={goMenu}>Main Menu</button>
        </div>
      )}

      {level.rendered}

      <div
        className="player"
        style={getPlayerPosition()}
      />

      <div
        className="exitDoor"
        style={getMapPosition(level.escapeDoor[1], level.escapeDoor[0], 13 / 15)}
      />

      {!!level.unlock.length && (
        <div
          key={`unlock-${levelNumber}`}
          className="unlock"
          style={getMapPosition(level.unlock[0], level.unlock[1], 2 / 3)}
        />
      )}

      {!!(level.lock.length && locked) && (
        <div
          key={`lock-${levelNumber}`}
          className="lock"
          style={getMapPosition(level.lock[0], level.lock[1], 1)}
        />
      )}

      <div className="gameFooter">
        <h3 className="info alignLeft">Time left: {timeLeft}</h3>
        <h3 className="info alignMiddle">Score: {score}</h3>
        <h3 className="info alignRight">Moves left: {movesLeft}</h3>
      </div>
    </div>
  );
};

Play.propTypes = {
  levelNumber: PropTypes.number.isRequired,
  score: PropTypes.number.isRequired,
  setLevelNumber: PropTypes.func.isRequired,
  setScore: PropTypes.func.isRequired,
  setScreen: PropTypes.func.isRequired
}

export default React.memo(Play);
