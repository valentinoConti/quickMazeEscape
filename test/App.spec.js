import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import App from '../src/App';
import Menu from '../src/screens/Menu'
import Play from '../src/screens/Play'

configure({adapter: new Adapter()});

describe('Main App tests', () => {
  let wrapper;

  it('should take/compare App snapshot', () => {
    wrapper = mount(
      <App />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should test the screen changing functionality', () => {
    wrapper = mount(
      <App />
    );
    expect(wrapper.find(Menu).exists()).toBeTruthy();
    expect(wrapper.find(Play).exists()).not.toBeTruthy();

    act(() => {
      wrapper.find(Menu).props().setScreen('play');
    });
    wrapper.update();

    expect(wrapper.find(Menu).exists()).not.toBeTruthy();
    expect(wrapper.find(Play).exists()).toBeTruthy();

  });

  it('should test nextLevel functionality', () => {
    const firstKey = wrapper.find(Play).key();

    act(() => {
      wrapper.find(Play).props().setLevelNumber(2);
    });
    wrapper.update();

    const secondKey = wrapper.find(Play).key();

    expect(firstKey).not.toContain(secondKey);
  })
});