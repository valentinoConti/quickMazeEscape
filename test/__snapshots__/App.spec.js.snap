// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main App tests should take/compare App snapshot 1`] = `
<App>
  <Menu
    setScreen={[Function]}
  >
    <div
      id="menuScreen"
    >
      <h1
        className="menuTitle"
      >
        Quick-Maze Escape
      </h1>
      <button
        className="menuButton"
        id="playButton"
        onClick={[Function]}
      >
        Play
      </button>
      <button
        className="menuButton marginTop"
        id="instructionsButton"
        onClick={[Function]}
      >
        Instructions
      </button>
      <button
        className="menuButton marginTop"
        id="fullscreenButton"
        onClick={[Function]}
      >
        Fullscreen
      </button>
      <footer>
        Valentino Conti
      </footer>
    </div>
  </Menu>
</App>
`;
